### PR TITLE
feat: backport guild forum support to v13

### PR DESCRIPTION
**Pull request by aiko-chan-ai**
**Please describe the changes this PR makes and why it should be merged:**

image

https://github.com/discordjs/discord.js/pull/7791

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by Jiralite** [comment]
> This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

What's the breaking change?

File: src/structures/ThreadChannel.js [hunk 1]
```diff
@@ -419,6 +440,16 @@ class ThreadChannel extends Channel {
     return this.edit({ rateLimitPerUser }, reason);
   }
 
+  /**
+   * Set the applied tags for this channel (only applicable to forum threads)
+   * @param {GuildForumTag[]} appliedTags The tags to set for this channel
```
Comment (jaw0r3k) [inline src/structures/ThreadChannel.js:445]:
```suggestion
   * @param {Snowflake[]} appliedTags The tags to set for this channel
```
as some pr says api takes id as it
also change typings 
or from my suggestion u can use somehow resolver

File: typings/index.d.ts [hunk 2]
```diff
@@ -2526,6 +2578,7 @@ export class ThreadChannel extends TextBasedChannelMixin(Channel, ['fetchWebhook
   public setInvitable(invitable?: boolean, reason?: string): Promise<ThreadChannel>;
   public setLocked(locked?: boolean, reason?: string): Promise<ThreadChannel>;
   public setName(name: string, reason?: string): Promise<ThreadChannel>;
+  public setAppliedTags(appliedTags: GuildForumTag[], reason?: string): Promise<ThreadChannel>;
```
Comment (jaw0r3k) [inline typings/index.d.ts:2581]:
```suggestion
  public setAppliedTags(appliedTags: Snowflake[], reason?: string): Promise<ThreadChannel>;
```

**Comment by jaw0r3k** [comment]
also can u add
#8645, #8643, #8638

File: src/structures/ThreadChannel.js (hunk 1 above)
Comment (aiko-chan-ai) [reply src/structures/ThreadChannel.js:445]: image

Comment (jaw0r3k) [reply src/structures/ThreadChannel.js:445]: > image

It is a array of object
but appliedTags takes ids
image

Comment (aiko-chan-ai) [reply src/structures/ThreadChannel.js:445]: > It is a array of objects but appliedTags takes ids image

oh

**Comment by aiko-chan-ai** [comment]
> > This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
> 
> What's the breaking change? If there is one, this cannot be merged.

probably not :c

File: src/managers/GuildForumThreadManager.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  /**
+   * Creates a new thread in the channel.
+   * @param {GuildForumThreadCreateOptions} [options] Options to create a new thread
+   * @returns {Promise<ThreadChannel>}
+   * @example
+   * // Create a new forum post
+   * forum.threads
+   *   .create({
+   *     name: 'Food Talk',
+   *     autoArchiveDuration: ThreadAutoArchiveDuration.OneHour,
```
Comment (jaw0r3k) [inline src/managers/GuildForumThreadManager.js:41]:
```suggestion
   *     autoArchiveDuration: 60,
```
OneHour iss for v14

**Review by jaw0r3k — CHANGES_REQUESTED** [review]

File: src/managers/GuildForumThreadManager.js (hunk 3 above)
Comment (aiko-chan-ai) [reply src/managers/GuildForumThreadManager.js:41]: > OneHour is for v14

oki :c

Comment (jaw0r3k) [reply src/managers/GuildForumThreadManager.js]: why u cant just commit suggestion?

Comment (aiko-chan-ai) [reply src/managers/GuildForumThreadManager.js]: > why u cant just commit suggestion?

I'm using vscode, so I think it's more convenient :v

Comment (jaw0r3k) [reply src/managers/GuildForumThreadManager.js]: > > why u cant just commit suggestion?
> 
> I'm using vscode, so I think it's more convenient :v

i think its just simpler to click it one time
that going to vscode and editing it

**Review by jaw0r3k — APPROVED** [review]

**Comment by JJRcop** [comment]
> probably not :c

Can you be more clear on this? What is the breaking change?

A "breaking change" means the modifications will cause old code written for the previous version to not function the same or create errors. Does this pull request have any of that? Please tell us!

**Comment by aiko-chan-ai** [comment]
> > probably not :c
> 
> Can you be more clear on this? What is the breaking change?
> 
> A "breaking change" means the modifications will cause old code written for the previous version to not function the same or create errors. Does this pull request have any of that? Please tell us!

i changed the name of the Class `ThreadManager`, same as v14. Still working the same way

**Comment by jaw0r3k** [comment]
> > Can you be more clear on this? What is the breaking change?
> > A "breaking change" means the modifications will cause old code written for the previous version to not function the same or create errors. Does this pull request have any of that? Please tell us!
> 
> i changed the name of the Class `ThreadManager`, same as v14. Still working the same way

name is the same 
just there is more managers

**Comment by aiko-chan-ai** [comment]
> name is the same just there is more managers

yep

**Comment by aiko-chan-ai** [comment]
> > Can you be more clear on this? What is the breaking change?
> > A "breaking change" means the modifications will cause old code written for the previous version to not function the same or create errors. Does this pull request have any of that? Please tell us!
> 
> i changed the name of the Class `ThreadManager`, same as v14. Still working the same way

So it doesn't have `breaking change` :v

**Comment by Jiralite** [comment]
Make sure you get #8662 in here too!

**Comment by jaw0r3k** [comment]
And #8661

**Comment by jaw0r3k** [comment]
Also #8664

**Comment by aiko-chan-ai** [comment]
done

File: src/managers/GuildChannelManager.js [hunk 4]
```diff
@@ -282,6 +284,7 @@ class GuildChannelManager extends CachedManager {
         rate_limit_per_user: data.rateLimitPerUser,
         default_auto_archive_duration: defaultAutoArchiveDuration,
         permission_overwrites,
+        tags: 'flags' in data ? ChannelFlags.resolve(data.flags) : undefined,
```
Comment (jaw0r3k) [inline src/managers/GuildChannelManager.js:287]:
```suggest
        fags: 'flags' in data ? ChannelFlags.resolve(data.flags) : undefined,
```
it should be flags

File: typings/index.d.ts [hunk 5]
```diff
@@ -4229,6 +4229,7 @@ export interface ChannelData {
   availableTags?: GuildForumTagData[];
   defaultReactionEmoji?: DefaultReactionEmoji;
   defaultThreadRateLimitPerUser?: number;
+  tags?: ChannelFlagsResolvable;
```
Comment (jaw0r3k) [inline typings/index.d.ts:4232]:
```suggestion
  fags?: ChannelFlagsResolvable;
```
also there

File: src/managers/GuildChannelManager.js (hunk 4 above)
Comment (thehairy) [reply src/managers/GuildChannelManager.js:287]: Please, don't misspell flags like this.

File: typings/index.d.ts (hunk 5 above)
Comment (thehairy) [reply typings/index.d.ts:4232]: Please, don't misspell flags like this.

Comment (jaw0r3k) [reply typings/index.d.ts:4232]: ah yeah lol

**Comment by aiko-chan-ai** [comment]
@jaw0r3k https://github.com/discordjs/discord.js/pull/8637/files hmm...

**Comment by jaw0r3k** [comment]
> @jaw0r3k https://github.com/discordjs/discord.js/pull/8637/files hmm...

yeah i fixed it
> with more typos but xD

**Review by jaw0r3k — APPROVED** [review]

**Comment by jaw0r3k** [comment]
And #8677

File: src/structures/ForumChannel.js [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+/**
+ * @typedef {Object} DefaultReactionEmoji
+ * @property {?Snowflake} id The id of a guild's custom emoji
+ * @property {?string} name The unicode character of the emoji
+ */
+
+/**
+ * Represents a channel that only contains threads
+ * @extends {GuildChannel}
```
Comment (jaw0r3k) [inline src/structures/ForumChannel.js:39]:
```suggestion
 * @extends {GuildChannel}
 * @implements {TextBasedChannel}
```

File: package.json [hunk 7]
```diff
@@ -56,7 +56,7 @@
     "@sapphire/async-queue": "^1.5.0",
     "@types/node-fetch": "^2.6.2",
     "@types/ws": "^8.5.3",
-    "discord-api-types": "^0.33.5",
+    "discord-api-types": "^0.37.11",
```
Comment (vladfrangu) [inline package.json:59]: Won't this cause a version clash with the older version of /builders?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: package.json (hunk 7 above)
Comment (iCrawl) [reply package.json:59]: I guess we have to bring in those types manually 😓

Comment (aiko-chan-ai) [reply package.json:59]: @vladfrangu I need a newer version to use `SortOrderType`

Comment (aiko-chan-ai) [reply package.json]: https://github.com/discordjs/discord.js/pull/8651/commits/54267f584a4e38cb58421cf7d26785a5a621dfcd
oki 👍

**Review by vladfrangu — APPROVED** [review]

File: src/managers/GuildForumThreadManager.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+      messagePayload = message.resolveData();
+    } else {
+      messagePayload = MessagePayload.create(this, message).resolveData();
+    }
+
+    const { data: body, files } = await messagePayload.resolveFiles();
+
+    if (autoArchiveDuration === 'MAX') autoArchiveDuration = resolveAutoArchiveMaxLimit(this.channel.guild);
+
+    const data = await path.threads.post({
```
Comment (NotSugden) [inline src/managers/GuildForumThreadManager.js:76]:
```suggestion
    if (!message) {
      throw new TypeError('GUILD_FORUM_MESSAGE_REQUIRED');
    }

    let messagePayload;

    if (message instanceof MessagePayload) {
      messagePayload = message.resolveData();
    } else {
      messagePayload = MessagePayload.create(this, message).resolveData();
    }

    const { data: body, files } = await messagePayload.resolveFiles();

    if (autoArchiveDuration === 'MAX') autoArchiveDuration = resolveAutoArchiveMaxLimit(this.channel.guild);

    const data = await this.client.api.channels(this.channel.id).threads.post({
```
we don't re-assign path so theres no point making it a variable

is it possible we could also use a tenary operator to assign to `messagePayload` ?

File: src/structures/ForumChannel.js [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    if ('default_reaction_emoji' in data) {
+      /**
+       * The emoji to show in the add reaction button on a thread in a guild forum channel
+       * @type {?DefaultReactionEmoji}
+       */
+      this.defaultReactionEmoji = data.default_reaction_emoji
+        ? transformAPIGuildDefaultReaction(data.default_reaction_emoji)
+        : null;
```
Comment (NotSugden) [inline src/structures/ForumChannel.js:76]:
```suggestion
      this.defaultReactionEmoji = data.default_reaction_emoji && transformAPIGuildDefaultReaction(data.default_reaction_emoji);
```
we can do this because if `default_reaction_emoji` is null it will assign null, else it will transform it

File: src/structures/ForumChannel.js [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+    } else {
+      this.availableTags ??= [];
+    }
+
+    if ('default_reaction_emoji' in data) {
+      /**
+       * The emoji to show in the add reaction button on a thread in a guild forum channel
+       * @type {?DefaultReactionEmoji}
+       */
+      this.defaultReactionEmoji = data.default_reaction_emoji && transformAPIGuildDefaultReaction(data.default_reaction_emoji);
```
Comment (kyranet) [inline src/structures/ForumChannel.js:74]: Please format the code, there's a linter error here.

**Review by kyranet — DISMISSED** [review]

**Comment by nk980113** [comment]
when will this pr get merged?
i think this backport is completed well

**Comment by aiko-chan-ai** [comment]
> when will this pr get merged? i think this backport is completed well

idk

**Review by iCrawl — APPROVED** [review]

File: src/managers/GuildForumThreadManager.js [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+class GuildForumThreadManager extends ThreadManager {
+  /**
+   * The channel this Manager belongs to
+   * @name GuildForumThreadManager#channel
+   * @type {ForumChannel}
+   */
+
+  /**
+   * @typedef {BaseMessageOptions} GuildForumThreadMessageCreateOptions
+   * @property {stickers} [stickers] The stickers to send with the message
```
Comment (Jiralite) [inline src/managers/GuildForumThreadManager.js:21]: `stickers` is not a type. This isn't right.

**Review by Jiralite — CHANGES_REQUESTED** [review]

File: src/managers/GuildForumThreadManager.js [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+class GuildForumThreadManager extends ThreadManager {
+  /**
+   * The channel this Manager belongs to
+   * @name GuildForumThreadManager#channel
+   * @type {ForumChannel}
+   */
+
+  /**
+   * @typedef {BaseMessageOptions} GuildForumThreadMessageCreateOptions
+   * @property {Sticker} [stickers] The stickers to send with the message
```
Comment (Jiralite) [inline src/managers/GuildForumThreadManager.js:21]: Looking at this:

https://github.com/discordjs/discord.js/blob/v13/src/structures/MessagePayload.js#L210

It's not accurate still as it accepts even ids.

File: src/util/Constants.js [hunk 13]
```diff
@@ -1383,6 +1385,15 @@ exports.GuildScheduledEventEntityTypes = createEnum([null, 'STAGE_INSTANCE', 'VO
  */
 exports.VideoQualityModes = createEnum([null, 'AUTO', 'FULL']);
 
+/**
+ * Sort {@link ForumChannel} posts by ?
```
Comment (Jiralite) [inline src/util/Constants.js:1389]: By what? Maybe reword its general description.

**Comment by nk980113** [comment]
Nice
It can get merged
Looking forward to version 13.13

**Review by kyranet — APPROVED** [review]
As a general review, we should not use `resolveAutoArchiveMaxLimit`. We should also tweak the function (in this PR or another one) to always return the maximum value, as they're not longer boost-locked.

I can maybe accept `'MAX'` as a valid option because v13, but it should be discouraged *and* deprecated (possibly with a deprecation warning). This can be achieved by modifying the function itself, so you don't have to modify multiple files.

However, because that's outside of this PR's scope, I'm approving this anyways. Feel free to make a follow-up PR addressing this review, though.

**Comment by aiko-chan-ai** [comment]
hmm, do you mean the `SEVEN_DAY_THREAD_ARCHIVE` feature can be used on all servers?

**Comment by kyranet** [comment]
That's right — https://github.com/discord/discord-api-docs/pull/4825 — can't find a blog post or an ddevs announcement for this, but I just tried setting in a test server of mine (without boosts) to 7 days and it worked just fine.

File: src/structures/Message.js [hunk 14]
```diff
@@ -855,9 +867,9 @@ class Message extends Base {
    * archived. This can be:
    * * `60` (1 hour)
    * * `1440` (1 day)
-   * * `4320` (3 days) <warn>This is only available when the guild has the `THREE_DAY_THREAD_ARCHIVE` feature.</warn>
-   * * `10080` (7 days) <warn>This is only available when the guild has the `SEVEN_DAY_THREAD_ARCHIVE` feature.</warn>
-   * * `'MAX'` Based on the guild's features
+   * * `4320` (3 days)
+   * * `10080` (7 days)
+   * * `'MAX'` Based on the guild's features <warn>The `'MAX'` option is deprecated and will be removed in the next major version.</warn>
```
Comment (kyranet) [inline src/structures/Message.js:872]:
```suggestion
   * * `'MAX'` (7 days) <warn>This option is deprecated and will be removed in the next major version.</warn>
```

**Review by kyranet — COMMENTED** [review]
Just one small docs thing 👍🏼

**Comment by aiko-chan-ai** [comment]
> Just one small docs thing 👍🏼

I'm not good at writing docs 😢

**Comment by kyranet** [comment]
Last commit looks good, but as I said in the discord.js guild, you need to change the doc because `'MAX'` always yields 7 days and does not longer read guild features anymore.

**Comment by nk980113** [comment]
so when will it get merged?
